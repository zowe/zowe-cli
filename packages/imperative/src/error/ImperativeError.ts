/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IImperativeError } from "./doc/IImperativeError";
import { IImperativeErrorParms } from "./doc/IImperativeErrorParms";

/**
 *
 * @export
 * @class ImperativeError
 * @extends {Error}
 */
export class ImperativeError extends Error {
    /**
     * The message generated/specified for the error - used for display/message/diagnostic purposes
     * @private
     * @type {string}
     * @memberof ImperativeError
     */
    private mMessage: string;

    /**
     * Construct the Imperative error object. Contains the defaults of the error and optionally captures diagnostics
     * and other information.
     * @param {IImperativeError} details - the error details and text (stack, messages, etc.)
     * @param {IImperativeErrorParms} parms - control parameters to indicate logging of node-report and more
     */
    constructor(public mDetails: IImperativeError, parms?: IImperativeErrorParms) {
        super();
        this.mMessage = mDetails.msg;

        /**
         * If parms are present, handle them, otherwise perform the default diagnostic collection
         */
        if (parms) {
            /**
             * Append a tag if present
             */
            if (parms.tag) {
                this.mMessage = parms.tag + ": " + this.mMessage;
            }
        }
    }

    /**
     * Return causeErrors
     * @readonly
     * @type {any[]}
     * @memberof ImperativeError
     */
    public get causeErrors(): any {
        return this.mDetails.causeErrors;
    }

    /**
     * Return additionalDetails
     * @readonly
     * @type {string}
     * @memberof ImperativeError
     */
    public get additionalDetails(): string {
        return this.mDetails.additionalDetails;
    }

    /**
     * Return IImperativeError object
     * @readonly
     * @type {IImperativeError}
     * @memberof ImperativeError
     */
    public get details(): IImperativeError {
        return this.mDetails;
    }

    /**
     * Return errorCode
     * @readonly
     * @type {string}
     * @memberof ImperativeError
     */
    public get errorCode(): string {
        return this.mDetails.errorCode;
    }

    /**
     * Return whether or not the error dump should be suppressed
     * @readonly
     * @type {string}
     * @memberof ImperativeError
     */
    public get suppressDump(): boolean {
        return this.mDetails.suppressDump;
    }

    /**
     * Return stack info
     * @readonly
     * @type {string}
     * @memberof ImperativeError
     */
    public get stack(): string {
        return this.mDetails.stack;
    }

    /**
     * Accessor for the error message.
     * @readonly
     * @return {string}: The error message
     * @memberof ImperativeError
     */
    public get message(): string {
        return this.mMessage;
    }

}
