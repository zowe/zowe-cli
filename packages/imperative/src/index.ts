/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

export * from "./censor";
export * from "./cmd";
export * from "./config";
export * from "./console";
export * from "./constants";
export * from "./events";
export * from "./error";
export * from "./expect";
export * from "./imperative";
// export * from "./interfaces";
export * from "./io";
export * from "./logger";
export * from "./messages";
export * from "./operations";
export * from "./profiles";
export * from "./rest";
export * from "./security";
export * from "./settings";
export * from "./utilities";
