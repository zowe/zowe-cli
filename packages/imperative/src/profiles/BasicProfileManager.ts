/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { AbstractProfileManager } from "./abstract/AbstractProfileManager";
import {
    IDeleteProfile,
    ILoadProfile,
    IMetaProfile,
    IProfile,
    IProfileDeleted,
    IProfileInitialized,
    IProfileLoaded,
    IProfileManagerInit,
    IProfileSaved,
    IProfileTypeConfiguration,
    IProfileUpdated,
    IProfileValidated,
    ISaveProfile,
    IUpdateProfile,
    IValidateProfileWithSchema
} from "./doc";

import { ImperativeExpect } from "../expect";
import { isNullOrUndefined } from "util";
import { ImperativeError } from "../error";
import { ProfileIO } from "./utils";

/**
 * Basic Profile Manager is the most basic implementation of the Imperative Profile Manager. In general, it invokes
 * all of the utility/services from the Abstract Profile manager to load, save, delete, validate, etc. Imperative
 * profiles. See the "AbstractProfileManager" header for more profile management details.
 *
 * The main differences between the abstract and the basic include:
 *
 * 1) The "loadAll()" API in the basic profile manager loads ALL profiles from all types.
 * 2) The Basic Profile Manager includes the "initialize()" API, which will create all type directories and persist
 *    the schema in the meta files.
 *
 * The Basic Profile Manager can be used "stand-alone" from an Imperative CLI. The intent is to provide apps built
 * using Imperative CLI's to take advantage of the profiles that the user has defined, without having to "locate" the
 * configuration documents used to construct the CLI's. This is why the initialize() API persists the configuration
 * documents within the meta files for each type.
 *
 * @export
 * @class BasicProfileManager
 * @extends {AbstractProfileManager<T>}
 * @template T
 */
export class BasicProfileManager<T extends IProfileTypeConfiguration> extends AbstractProfileManager<T> {
    /**
     * Static method to initialize the profile environment. Accepts the profile root directory (normally supplied by
     * your Imperative configuration documents) and all profile "type" configuration documents and constructs the directories
     * needed to manage profiles of all types. You must execute this method before beginning to use profiles OR you must
     * supply all the type configuration documents (normally obtained from your Imperative configuration document) to
     * the constructor of
     * @static
     * @param {IProfileManagerInit} parms
     * @returns {Promise<IProfileInitialized[]>}
     * @memberof AbstractProfileManager
     */
    public static async initialize(parms: IProfileManagerInit): Promise<IProfileInitialized[]> {
        // Validate the input parameters - TODO: Validate all
        ImperativeExpect.toNotBeNullOrUndefined(
            parms,
            `A request was made to initialize the profile environment, but no parameters were supplied.`
        );
        ImperativeExpect.keysToBeDefined(parms, ["configuration"],
            `A request was made to initialize the profile environment, but no configuration documents were supplied.`
        );
        ImperativeExpect.keysToBeDefinedAndNonBlank(parms, ["profileRootDirectory"],
            `A request was made to initialize the profile environment, but the profile root directory was not supplied.`
        );
        ImperativeExpect.keysToBeAnArray(parms, true, ["configuration"],
            `A request was mad to initialize the profile environment, but the configuration provided is invalid (not an array or of length 0).`
        );

        // Set any defaults
        parms.reinitialize = (isNullOrUndefined(parms.reinitialize)) ? false : parms.reinitialize;

        // Create the profile root directory (if necessary)
        ProfileIO.createProfileDirs(parms.profileRootDirectory);

        // Iterate through the types and create this types configuration document - create a new instance of the
        // Manager to create the other types
        const responses: IProfileInitialized[] = [];
        for (const config of parms.configuration) {

            // Construct the profile type directory
            const profileTypeRootDir = parms.profileRootDirectory + "/" + config.type + "/";
            ProfileIO.createProfileDirs(profileTypeRootDir);

            // Meta file path and name
            const metaFilePath = profileTypeRootDir + config.type
                + AbstractProfileManager.META_FILE_SUFFIX + AbstractProfileManager.PROFILE_EXTENSION;

            // Construct the default meta file
            const defaultMetaFile: IMetaProfile<IProfileTypeConfiguration> = {
                defaultProfile: undefined,
                configuration: config
            };

            // If the directory doesn't exist, create it and the default meta file for this type
            // If the directory exists and re-init was specified, write out the default meta file
            // If it exists and re-init was not specified, leave it alone
            if (!ProfileIO.exists(metaFilePath)) {
                ProfileIO.writeMetaFile(defaultMetaFile, metaFilePath);
                responses.push({
                    message: `Profile environment initialized for type "${config.type}".`
                });
            } else if (parms.reinitialize) {
                ProfileIO.writeMetaFile(defaultMetaFile, metaFilePath);
                responses.push({
                    message: `Profile environment re-initialized for type "${config.type}".`
                });
            }
        }

        return responses;
    }

    /**
     * Loads all profiles from every type. Profile types are deteremined by reading all directories within the
     * profile root directory.
     * @returns {Promise<IProfileLoaded[]>} - The list of all profiles for every type
     * @memberof BasicProfileManager
     */
    public async loadAll(): Promise<IProfileLoaded[]> {
        this.log.trace(`Loading all profiles...`);
        // Load all the other profiles for other types
        const loadTheirProfiles: any[] = [];
        let loadList: string = "";
        for (const typeConfig of this.profileTypeConfigurations) {
            const typeProfileManager = new BasicProfileManager({
                profileRootDirectory: this.profileRootDirectory,
                typeConfigurations: this.profileTypeConfigurations,
                type: typeConfig.type,
                logger: this.log,
                loadCounter: this.loadCounter
            });

            // Get all the profile names for the type and attempt to load every one
            const names: string[] = typeProfileManager.getAllProfileNames();
            for (const name of names) {
                this.log.debug(`Loading profile "${name}" of type "${typeConfig.type}".`);
                loadTheirProfiles.push(typeProfileManager.load({name, failNotFound: true, loadDependencies: false}));
                loadList += `\nLoading "${name}" of type "${typeConfig.type}"`;
            }
        }

        // Construct the full list for return
        let allProfiles: IProfileLoaded[] = [];
        try {
            this.log.trace(`Awaiting all loads...`);
            const theirProfiles = await Promise.all(loadTheirProfiles);
            for (const theirs of theirProfiles) {
                allProfiles = allProfiles.concat(theirs);
            }
            this.log.trace(`All loads complete.`);
        } catch (e) {
            const msg: string = `An error occurred attempting to load all profiles of every type. ` +
                `Load List: ${loadList}\nError Details: "${e.message}"`;
            this.log.error(msg);
            throw new ImperativeError({msg});
        }

        return allProfiles;
    }

    /**
     * Loads all dependencies for the profile specified - returns the list in the response structure. Sub-dependencies
     * are also loaded.
     * @protected
     * @param {string} name - the name of hte profile to load dependencies for
     * @param {IProfile} profile - The profile to load dependencies.
     * @param {boolean} [failNotFound=true] - Indicates that you want to avoid failing the request for "not found" errors.
     * @returns {Promise<IProfileLoaded[]>} - The list of profiles loaded with all dependencies.
     * @memberof BasicProfileManager
     */
    protected loadDependencies(name: string, profile: IProfile, failNotFound = true): Promise<IProfileLoaded[]> {
        return new Promise<IProfileLoaded[]>((dependenciesLoaded, loadFailed) => {

            // Construct a list of promises to load all profiles
            const promises: Array<Promise<IProfileLoaded>> = [];
            const responses: IProfileLoaded[] = [];
            if (!isNullOrUndefined(profile.dependencies)) {
                this.log.debug(`Loading dependencies for profile of "${this.profileType}".`);
                let list: string = "";
                for (const dependency of profile.dependencies) {
                    this.log.debug(`Loading dependency "${dependency.name}" of type "${dependency.type}".`);
                    promises.push(new BasicProfileManager({
                        profileRootDirectory: this.profileRootDirectory,
                        typeConfigurations: this.profileTypeConfigurations,
                        type: dependency.type,
                        logger: this.log,
                        loadCounter: this.loadCounter
                    }).load({name: dependency.name, failNotFound}));
                    list += `\nType: "${dependency.type}" Name: "${dependency.name}"`;
                }

                // Wait for all the promises to complete
                Promise.all(promises).then((loadResponses) => {
                    this.log.debug(`All dependencies loaded for profile of type "${this.profileType}".`);
                    // Append the responses for return to caller
                    for (const response of loadResponses) {
                        responses.push(response);
                    }
                    dependenciesLoaded(responses);
                }).catch((loadsFailed) => {
                    this.log.error(`Failure to load dependencies for profile of type "${this.profileType}". ` +
                        `Details: ${loadsFailed.message}`);
                    const err: string = `An error occurred while loading the dependencies of profile ` +
                        `"${name}" of type "${this.profileType}". Dependency load list: ${list}\n\nError Details: ${loadsFailed.message}`;
                    loadFailed(new ImperativeError({msg: err, additionalDetails: loadsFailed}));
                });
            } else {
                this.log.trace(`Profile of type "${this.profileType}" has no dependencies.`);
                dependenciesLoaded([]);
            }
        });
    }

    /**
     * Save the profile to disk. First ensures that all dependencies are valid and writes the profile.
     * @protected
     * @param {ISaveProfile} parms - Save control params - see the interface for full details
     * @returns {Promise<IProfileSaved>} - Promise that is fulfilled when complete (or rejected with an Imperative Error)
     * @memberof BasicProfileManager
     */
    protected async saveProfile(parms: ISaveProfile): Promise<IProfileSaved> {
        // Validate that the dependencies listed exist before saving
        try {
            this.log.debug(`Loading dependencies for profile "${parms.name}" of type "${this.profileType}", ` +
                `checking if if they are valid (before save.)`);
            await this.loadDependencies(parms.name, parms.profile);
        } catch (e) {
            throw new ImperativeError({
                msg: `Could not save the profile, because one or more dependencies is invalid or does not exist.\n` +
                    `Load Error Details: ${e.message}`
            });
        }

        // Construct the full file path, write to disk, and return the response
        this.log.info(`Saving profile "${parms.name}" of type "${this.profileType}"...`);
        const path = this.constructFullProfilePath(parms.name);
        ProfileIO.writeProfile(path, parms.profile);
        this.log.info(`Profile "${parms.name}" of type "${this.profileType}" saved.`);
        return {
            path,
            overwritten: parms.overwrite || false,
            message: `Profile ("${parms.name}" of type "${this.profileType}") ` +
                `successfully written: ${path}`,
            profile: parms.profile
        };
    }

    /**
     * Load a profile from disk - invokes the "loadSpecificProfile" method in the abstract to perform the load.
     * @protected
     * @param {ILoadProfile} parms - Load control params - see the interface for full details
     * @returns {Promise<IProfileSaved>} - Promise that is fulfilled when complete (or rejected with an Imperative Error)
     * @memberof BasicProfileManager
     */
    protected async loadProfile(parms: ILoadProfile): Promise<IProfileLoaded> {
        const loadName: string = (parms.loadDefault || false) ? this.getDefaultProfileName() : parms.name;
        this.log.debug(`Loading profile "${loadName}" (load default: "${parms.loadDefault}") of type "${this.profileType}".`);
        return this.loadSpecificProfile(loadName, parms.failNotFound, parms.loadDependencies);
    }

    /**
     * Delete a profile from disk - invokes the "deleteProfileFromDisk" method in the abstract to perform the load.
     * @protected
     * @param {IDeleteProfile} parms - Delete control params - see the interface for full details
     * @returns {Promise<IProfileDeleted>} - Promise that is fulfilled when complete (or rejected with an Imperative Error)
     * @memberof BasicProfileManager
     */
    protected async deleteProfile(parms: IDeleteProfile): Promise<IProfileDeleted> {
        this.log.trace(`Removing profile "${parms.name}" of type "${this.profileType}".`);
        const path = this.deleteProfileFromDisk(parms.name);
        this.log.debug(`Profile "${parms.name}" of type "${this.profileType}" successfully deleted.`);
        return {
            path,
            message: `Profile "${parms.name}" of type "${this.profileType}" deleted successfully.`
        };
    }

    /**
     * Validate profile - ensures that the profile is valid agaisnt the schema and configuration document
     * @protected
     * @param {IValidateProfileWithSchema} parms - Validate control params - see the interface for full details
     * @returns {Promise<IProfileValidated>} - Promise that is fulfilled when complete (or rejected with an Imperative Error)
     * @memberof BasicProfileManager
     */
    protected async validateProfile(parms: IValidateProfileWithSchema): Promise<IProfileValidated> {
        this.log.trace(`Validating profile "${parms.name}" of type "${this.profileType}"`);
        // Ensure that the profile is not empty
        if (this.isProfileEmpty(parms.profile)) {
            throw new ImperativeError({
                msg: `The profile passed (name "${parms.name}" of type ` +
                    `"${this.profileType}") does not contain any content.`
            });
        }

        // If the configuration indicates this profile type has required dependencies, ensure that those are specified
        // on the profile object passed.
        this.validateRequiredDependenciesAreSpecified(parms.profile);

        // Validate the profile agaisnt the schema
        this.validateProfileAgainstSchema(parms.name, parms.profile, parms.strict);

        // Return the response
        this.log.debug(`Profile "${parms.name}" of type "${this.profileType}" is valid.`);
        return {
            message: `Profile "${parms.name}" of type "${this.profileType}" is valid.`
        };
    }

    /**
     * Update a profile - Accepts the "new" version of the profile and overwrites the existing profile on disk.
     * @protected
     * @param {IUpdateProfile} parms - Update control params - see the interface for full details
     * @returns {Promise<IProfileUpdated>} - Promise that is fulfilled when complete (or rejected with an Imperative Error)
     * @memberof BasicProfileManager
     */
    protected async updateProfile(parms: IUpdateProfile): Promise<IProfileUpdated> {
        this.log.trace(`Saving (as part of updating) profile "${parms.name}" of type "${this.profileType}".`);
        if (parms.merge) {
            this.log.debug(`Profile merging was requested. Loading the old version of the profile ` +
                `"${parms.name}" of type "${this.profileType}".`);
            const oldProfileLoad = await this.load({name: parms.name, failNotFound: true});
            parms.profile = this.mergeProfiles(oldProfileLoad.profile, parms.profile);
            this.log.debug(`Merged profile "${parms.name}" of type "${this.profileType}" with old version`);
        }
        const response = await this.save({
            name: parms.name,
            type: this.profileType,
            profile: parms.profile,
            overwrite: true
        });
        this.log.trace(`Save of profile "${parms.name}" of type "${this.profileType}" complete.`);
        return {
            path: response.path,
            message: `Profile "${parms.name}" of type "${this.profileType}" updated successfully.`,
            profile: response.profile
        };
    }
}
