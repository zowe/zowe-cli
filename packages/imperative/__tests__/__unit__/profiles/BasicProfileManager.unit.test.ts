/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

jest.mock("../../../src/utils/ProfileIO");
import { ImperativeError } from "../../../src/error/ImperativeError";
import { TestLogger } from "../../../__tests__/src/TestLogger";
import { IProfileLoaded } from "../../../src/profiles/doc/response/IProfileLoaded";
import {
    APPLE_PROFILE_TYPE,
    APPLE_TWO_REQ_DEP_BANANA_AND_STRAWBERRIES,
    BLUEBERRY_PROFILE_TYPE,
    FRUIT_BASKET_DIR,
    ONLY_APPLE,
    ONLY_BLUEBERRY,
    STRAWBERRY_AND_APPLE_NO_DEP,
    TEST_PROFILE_ROOT_DIR
} from "./TestConstants";
import { BasicProfileManager } from "../../../src/profiles/BasicProfileManager";
// UnitTestUtils.replaceIt();

describe("Basic Profile Manager", () => {
    it("should detect no parms when instantiating", () => {
        let error;
        try {
            const prof = new BasicProfileManager(undefined);
        } catch (e) {
            error = e;
            TestLogger.info(error);
        }
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toMatchSnapshot();
    });

    it("should detect that the profile directory is undefined", () => {
        let error;
        try {
            const prof = new BasicProfileManager({
                profileRootDirectory: undefined,
                typeConfigurations: ONLY_APPLE,
                type: APPLE_PROFILE_TYPE,
                logger: TestLogger.getTestLogger()
            });
        } catch (e) {
            error = e;
            TestLogger.info(error);
        }
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toMatchSnapshot();
    });

    it("should detect that the profile directory is blank", () => {
        let error;
        try {
            const prof = new BasicProfileManager({
                profileRootDirectory: " ",
                typeConfigurations: ONLY_APPLE,
                type: APPLE_PROFILE_TYPE,
                logger: TestLogger.getTestLogger()
            });
        } catch (e) {
            error = e;
            TestLogger.info(error);
        }
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toMatchSnapshot();
    });

    it("should detect that no type configuration is supplied", () => {
        let error;
        try {
            const prof = new BasicProfileManager({
                profileRootDirectory: TEST_PROFILE_ROOT_DIR,
                typeConfigurations: undefined,
                type: APPLE_PROFILE_TYPE,
                logger: TestLogger.getTestLogger()
            });
        } catch (e) {
            error = e;
            TestLogger.info(error);
        }
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toContain("An error occurred collecting all configurations from the profile root directory");
    });

    it("should detect that the type configuration is an empty array", () => {
        let error;
        try {
            const prof = new BasicProfileManager({
                profileRootDirectory: TEST_PROFILE_ROOT_DIR,
                typeConfigurations: [],
                type: APPLE_PROFILE_TYPE,
                logger: TestLogger.getTestLogger()
            });
        } catch (e) {
            error = e;
            TestLogger.info(error);
        }
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toContain("An error occurred collecting all configurations from the profile root directory");
    });

    it("should detect if the type is undefined", () => {
        let error;
        try {
            const prof = new BasicProfileManager({
                profileRootDirectory: TEST_PROFILE_ROOT_DIR,
                typeConfigurations: ONLY_APPLE,
                type: undefined,
                logger: TestLogger.getTestLogger()
            });
        } catch (e) {
            error = e;
            TestLogger.info(error);
        }
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toMatchSnapshot();
    });

    it("should detect if the type is blank", () => {
        let error;
        try {
            const prof = new BasicProfileManager({
                profileRootDirectory: TEST_PROFILE_ROOT_DIR,
                typeConfigurations: ONLY_APPLE,
                type: " ",
                logger: TestLogger.getTestLogger()
            });
        } catch (e) {
            error = e;
            TestLogger.info(error);
        }
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toMatchSnapshot();
    });

    it("should detect that a type not found within the configurations", () => {
        let error;
        try {
            const prof = new BasicProfileManager({
                profileRootDirectory: TEST_PROFILE_ROOT_DIR,
                typeConfigurations: ONLY_APPLE,
                type: "bad_apple",
                logger: TestLogger.getTestLogger()
            });
        } catch (e) {
            error = e;
            TestLogger.info(error);
        }
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toContain(
            "Expect Error: Could not locate the profile type configuration for \"bad_apple\" within the input configuration list passed."
        );
    });

    it("should allow us to instantiate the cli profile manager", () => {
        let error;
        try {
            const prof = new BasicProfileManager({
                profileRootDirectory: TEST_PROFILE_ROOT_DIR,
                typeConfigurations: ONLY_APPLE,
                type: APPLE_PROFILE_TYPE,
                logger: TestLogger.getTestLogger()
            });
            TestLogger.info("Profile Manager Created");
        } catch (e) {
            error = e;
            TestLogger.error(e);
        }
        expect(error).toBeUndefined();
    });

    it("should load all profiles", async () => {
        const prof = new BasicProfileManager({
            profileRootDirectory: TEST_PROFILE_ROOT_DIR,
            typeConfigurations: STRAWBERRY_AND_APPLE_NO_DEP,
            type: APPLE_PROFILE_TYPE,
            logger: TestLogger.getTestLogger()
        });

        let error;
        let responses: IProfileLoaded[];
        try {
            responses = await prof.loadAll();
        } catch (e) {
            error = e;
            TestLogger.error(e);
        }
        expect(error).toBeUndefined();
        expect(responses).toMatchSnapshot();
    });

    it("should detect ill formed profiles during a load all", async () => {
        const prof = new BasicProfileManager({
            profileRootDirectory: TEST_PROFILE_ROOT_DIR,
            typeConfigurations: APPLE_TWO_REQ_DEP_BANANA_AND_STRAWBERRIES,
            type: APPLE_PROFILE_TYPE,
            logger: TestLogger.getTestLogger()
        });

        let error;
        let responses: IProfileLoaded[];
        try {
            responses = await prof.loadAll();
        } catch (e) {
            error = e;
            TestLogger.error(e);
        }
        expect(error).toBeDefined();
        expect(error.message).toMatchSnapshot();
    });

    it("should create an instance and read all configurations from the meta files", async () => {
        const prof = new BasicProfileManager({
            profileRootDirectory: TEST_PROFILE_ROOT_DIR + FRUIT_BASKET_DIR,
            type: APPLE_PROFILE_TYPE,
            logger: TestLogger.getTestLogger()
        });
        expect(prof.configurations).toMatchSnapshot();
    });

    it("should fail a create if no configurations and passed and none can be read from disk", async () => {
        let error;
        try {
            const prof = new BasicProfileManager({
                profileRootDirectory: TEST_PROFILE_ROOT_DIR,
                type: APPLE_PROFILE_TYPE,
                logger: TestLogger.getTestLogger()
            });
        } catch (e) {
            error = e;
        }
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toMatchSnapshot();
    });

    it("should allow us to set the default in the meta profile", () => {
        let error;
        let response;
        try {
            const prof = new BasicProfileManager({
                profileRootDirectory: TEST_PROFILE_ROOT_DIR,
                typeConfigurations: ONLY_BLUEBERRY,
                type: BLUEBERRY_PROFILE_TYPE,
                logger: TestLogger.getTestLogger()
            });

            response = prof.setDefault("tart_blueberry");
        } catch (e) {
            error = e;
            TestLogger.error(e);
        }
        expect(error).toBeUndefined();
        expect(response).toMatchSnapshot();
    });

    it("should fail a request to set the default if the profile is not found", () => {
        let error;
        let response;
        try {
            const prof = new BasicProfileManager({
                profileRootDirectory: TEST_PROFILE_ROOT_DIR,
                typeConfigurations: ONLY_BLUEBERRY,
                type: BLUEBERRY_PROFILE_TYPE,
                logger: TestLogger.getTestLogger()
            });
            response = prof.setDefault("bad_blueberry");
        } catch (e) {
            error = e;
            TestLogger.info(e);
        }
        expect(error).toBeDefined();
        expect(error instanceof ImperativeError).toBe(true);
        expect(error.message).toMatchSnapshot();
        expect(response).toBeUndefined();
    });
});
