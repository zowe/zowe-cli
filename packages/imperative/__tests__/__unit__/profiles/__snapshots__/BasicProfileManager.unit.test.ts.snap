// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Profile Manager should allow a re-initialize of the environment 1`] = `
Array [
  Object {
    "message": "Profile environment re-initialized for type \\"apple\\".",
  },
  Object {
    "message": "Profile environment re-initialized for type \\"strawberry\\".",
  },
  Object {
    "message": "Profile environment initialized for type \\"banana\\".",
  },
  Object {
    "message": "Profile environment re-initialized for type \\"grape\\".",
  },
]
`;

exports[`Basic Profile Manager should allow us to set the default in the meta profile 1`] = `"Default profile for type \\"blueberry\\" set to \\"tart_blueberry\\"."`;

exports[`Basic Profile Manager should create an instance and read all configurations from the meta files 1`] = `
Array [
  Object {
    "schema": Object {
      "description": "The simple apple configuration",
      "properties": Object {
        "age": Object {
          "type": "number",
        },
        "description": Object {
          "type": "string",
        },
        "rotten": Object {
          "type": "boolean",
        },
      },
      "required": Array [
        "description",
        "rotten",
        "age",
      ],
      "title": "The simple apple configuration",
      "type": "object",
    },
    "type": "apple",
  },
  Object {
    "schema": Object {
      "description": "The simple strawberry configuration",
      "properties": Object {
        "amount": Object {
          "type": "number",
        },
        "description": Object {
          "type": "string",
        },
      },
      "required": Array [
        "description",
        "amount",
      ],
      "title": "The simple strawberry configuration",
      "type": "object",
    },
    "type": "strawberry",
  },
  Object {
    "schema": Object {
      "description": "The simple banana configuration",
      "properties": Object {
        "bundle": Object {
          "type": "boolean",
        },
        "description": Object {
          "type": "string",
        },
      },
      "required": Array [
        "bundle",
      ],
      "title": "The simple banana configuration",
      "type": "object",
    },
    "type": "banana",
  },
  Object {
    "schema": Object {
      "description": "The simple banana configuration",
      "properties": Object {
        "color": Object {
          "type": "string",
        },
        "description": Object {
          "type": "string",
        },
      },
      "required": Array [
        "description",
        "color",
      ],
      "title": "The simple banana configuration",
      "type": "object",
    },
    "type": "grape",
  },
]
`;

exports[`Basic Profile Manager should detect blank profile directory on initialize 1`] = `"Expect Error: A request was made to initialize the profile environment, but the profile root directory was not supplied."`;

exports[`Basic Profile Manager should detect if the type is blank 1`] = `"Expect Error: No profile type supplied on the profile manager parameters."`;

exports[`Basic Profile Manager should detect if the type is undefined 1`] = `"Expect Error: No profile type supplied on the profile manager parameters."`;

exports[`Basic Profile Manager should detect ill formed profiles during a load all 1`] = `
"An error occurred attempting to load all profiles of every type. Load List: 
Loading \\"good_apple\\" of type \\"apple\\"
Loading \\"tasty_apples\\" of type \\"apple\\"
Loading \\"strawberry_and_apple\\" of type \\"strawberry\\"
Error Details: \\"Profile validation error during load of profile \\"good_apple\\" of type \\"apple\\". Error Details: Profile type \\"apple\\" specifies a required dependency of type \\"strawberry\\" on the \\"apple\\" profile type configuration document. A dependency of type \\"strawberry\\" was NOT listed on the input profile.\\""
`;

exports[`Basic Profile Manager should detect missing configuration on initialize 1`] = `"Expect Error: A request was made to initialize the profile environment, but no configuration documents were supplied."`;

exports[`Basic Profile Manager should detect missing parms on initialize 1`] = `"Expect Error: A request was made to initialize the profile environment, but no parameters were supplied."`;

exports[`Basic Profile Manager should detect missing profile directory on initialize 1`] = `"Expect Error: A request was made to initialize the profile environment, but the profile root directory was not supplied."`;

exports[`Basic Profile Manager should detect no parms when instantiating 1`] = `"Expect Error: Profile Manager input parms not supplied."`;

exports[`Basic Profile Manager should detect that the configuration passed is not an array 1`] = `"Expect Error: A request was mad to initialize the profile environment, but the configuration provided is invalid (not an array or of length 0)."`;

exports[`Basic Profile Manager should detect that the profile directory is blank 1`] = `"Expect Error: No profile root directory supplied on the profile manager parameters"`;

exports[`Basic Profile Manager should detect that the profile directory is undefined 1`] = `"Expect Error: No profile root directory supplied on the profile manager parameters"`;

exports[`Basic Profile Manager should fail a create if no configurations and passed and none can be read from disk 1`] = `"An error occurred collecting all configurations from the profile root directory \\"__tests__/__results__/test_profiles/root/dir/\\". Please supply the configurations on the profile manager constructor parameters OR initialize the profile manager environment. Details: No profile configurations found. Please initialize the profile manager OR supply the configurations to the profile manager."`;

exports[`Basic Profile Manager should fail a request to set the default if the profile is not found 1`] = `"Cannot update default profile for type \\"blueberry\\". The profile name specified (\\"bad_blueberry\\") does not exist. Please create before attempting to set the default."`;

exports[`Basic Profile Manager should initialize the environment 1`] = `
Array [
  Object {
    "message": "Profile environment initialized for type \\"apple\\".",
  },
  Object {
    "message": "Profile environment initialized for type \\"strawberry\\".",
  },
  Object {
    "message": "Profile environment initialized for type \\"banana\\".",
  },
  Object {
    "message": "Profile environment initialized for type \\"grape\\".",
  },
]
`;

exports[`Basic Profile Manager should load all profiles 1`] = `
Array [
  Object {
    "dependenciesLoaded": false,
    "dependencyLoadResponses": Array [],
    "failNotFound": true,
    "message": "Profile \\"good_apple\\" of type \\"apple\\" loaded successfully.",
    "name": "good_apple",
    "profile": Object {
      "age": 1,
      "description": "A tasty apple",
      "rotten": false,
    },
    "type": "apple",
  },
  Object {
    "dependenciesLoaded": false,
    "dependencyLoadResponses": Array [],
    "failNotFound": true,
    "message": "Profile \\"tasty_apples\\" of type \\"apple\\" loaded successfully.",
    "name": "tasty_apples",
    "profile": Object {
      "age": 1,
      "description": "tasty",
      "rotten": false,
    },
    "type": "apple",
  },
  Object {
    "dependenciesLoaded": false,
    "dependencyLoadResponses": Array [],
    "failNotFound": true,
    "message": "Profile \\"strawberry_and_apple\\" of type \\"strawberry\\" loaded successfully.",
    "name": "strawberry_and_apple",
    "profile": Object {
      "amount": 1000,
      "dependencies": Array [
        Object {
          "name": "good_apple",
          "type": "apple",
        },
      ],
      "description": "Tasty",
    },
    "type": "strawberry",
  },
]
`;

exports[`Basic Profile Manager should only initialize types not already defined in the environment 1`] = `
Array [
  Object {
    "message": "Profile environment initialized for type \\"banana\\".",
  },
]
`;
